File: iusj-school-service/src/main/java/com/example/iusj_school_service/services/SchoolService.java
package com.example.iusj_school_service.services;

import java.util.List;
import java.util.Optional;

import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.iusj_school_service.entities.Filiere;
import com.example.iusj_school_service.entities.School;
import com.example.iusj_school_service.repositories.FiliereRepository;
import com.example.iusj_school_service.repositories.SchoolRepository;

@Service
@Transactional
public class SchoolService {

    private final SchoolRepository repository;
    private final FiliereRepository filiereRepository;

    public SchoolService(SchoolRepository repository, FiliereRepository filiereRepository) {
        this.repository = repository;
        this.filiereRepository = filiereRepository;
    }

    public List<School> getAll(String name, School.Status status) {
        Specification<School> spec = SchoolSpecifications.withFilters(name, status);
        return repository.findAll(spec, Sort.by(Sort.Direction.ASC, "name"));
    }

    public Optional<School> getById(Long id) {
        return repository.findById(id);
    }

    public School create(School school) {
        // Set the school reference on each filiere before saving
        if (school.getFilieres() != null) {
            for (Filiere filiere : school.getFilieres()) {
                filiere.setSchool(school);
            }
        }
        return repository.save(school);
    }

    public Optional<School> update(Long id, School school) {
        return repository.findById(id).map(existing -> {
            school.setId(id);
            
            // Clear existing filieres and add updated ones
            existing.getFilieres().clear();
            
            if (school.getFilieres() != null) {
                for (Filiere filiere : school.getFilieres()) {
                    filiere.setSchool(existing);
                    existing.getFilieres().add(filiere);
                }
            }
            
            // Update other fields
            existing.setName(school.getName());
            existing.setCode(school.getCode());
            existing.setDescription(school.getDescription());
            existing.setAddress(school.getAddress());
            existing.setPhone(school.getPhone());
            existing.setEmail(school.getEmail());
            existing.setStatus(school.getStatus());
            
            return repository.save(existing);
        });
    }

    public void delete(Long id) {
        repository.deleteById(id);
    }

    public SchoolStats stats() {
        long total = repository.count();
        long active = repository.countByStatus(School.Status.ACTIVE);
        long inactive = repository.countByStatus(School.Status.INACTIVE);
        return new SchoolStats(total, active, inactive);
    }

    // === Filière methods ===

    public List<Filiere> getFilieresBySchoolId(Long schoolId) {
        return filiereRepository.findBySchoolId(schoolId);
    }

    public Optional<Filiere> getFiliereById(Long id) {
        return filiereRepository.findById(id);
    }

    public Filiere addFiliereToSchool(Long schoolId, Filiere filiere) {
        return repository.findById(schoolId).map(school -> {
            filiere.setSchool(school);
            return filiereRepository.save(filiere);
        }).orElseThrow(() -> new RuntimeException("School not found with id " + schoolId));
    }

    public Optional<Filiere> updateFiliere(Long filiereId, Filiere filiere) {
        return filiereRepository.findById(filiereId).map(existing -> {
            existing.setCode(filiere.getCode());
            existing.setNom(filiere.getNom());
            existing.setDescription(filiere.getDescription());
            existing.setStatus(filiere.getStatus());
            return filiereRepository.save(existing);
        });
    }

    public void deleteFiliere(Long filiereId) {
        filiereRepository.deleteById(filiereId);
    }

    public List<Filiere> getAllFilieres() {
        return filiereRepository.findAll(Sort.by(Sort.Direction.ASC, "code"));
    }

    public record SchoolStats(long total, long active, long inactive) {}
}

File: frontend/src/app/features/schools/services/school.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface School {
  id?: number;
  name: string;
  code?: string;
  description?: string;
  address?: string;
  phone?: string;
  email?: string;
  status: 'ACTIVE' | 'INACTIVE';
  filieres?: Filiere[];
}

export interface Filiere {
  id?: number;
  code: string;
  nom: string;
  description?: string;
  status: 'ACTIVE' | 'INACTIVE';
}

@Injectable({
  providedIn: 'root'
})
export class SchoolService {
  private apiUrl = 'http://localhost:8080/api/schools';

  constructor(private http: HttpClient) { }

  getAll(name?: string, status?: string): Observable<School[]> {
    let url = this.apiUrl;
    const params = new URLSearchParams();
    if (name) params.append('name', name);
    if (status) params.append('status', status);
    if (params.toString()) url += '?' + params.toString();
    return this.http.get<School[]>(url);
  }

  getById(id: number): Observable<School> {
    return this.http.get<School>(`${this.apiUrl}/${id}`);
  }

  create(school: School): Observable<School> {
    return this.http.post<School>(this.apiUrl, school);
  }

  update(id: number, school: School): Observable<School> {
    return this.http.put<School>(`${this.apiUrl}/${id}`, school);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Filière methods
  getAllFilieres(): Observable<Filiere[]> {
    return this.http.get<Filiere[]>(`${this.apiUrl}/filieres`);
  }

  getFilieresBySchool(schoolId: number): Observable<Filiere[]> {
    return this.http.get<Filiere[]>(`${this.apiUrl}/${schoolId}/filieres`);
  }

  addFiliere(schoolId: number, filiere: Filiere): Observable<Filiere> {
    return this.http.post<Filiere>(`${this.apiUrl}/${schoolId}/filieres`, filiere);
  }

  updateFiliere(id: number, filiere: Filiere): Observable<Filiere> {
    return this.http.put<Filiere>(`${this.apiUrl}/filieres/${id}`, filiere);
  }

  deleteFiliere(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/filieres/${id}`);
  }
}
